## You used two hidden layers. Try using one or three hidden layers, and see how doing so affects validation and test accuracy.

In [1]:
from keras.datasets import imdb

(train_data, train_labels), (test_data, test_labels) = imdb.load_data(
    num_words=10000)

Using TensorFlow backend.


In [4]:
import numpy as np

def vectorize_sequences(sequences, dimension=10000): 
    results = np.zeros((len(sequences), dimension)) 
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1. 
    return results

x_train = vectorize_sequences(train_data) 
x_test = vectorize_sequences(test_data)

y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')

In [5]:
x_val = x_train[:10000]
part_x_train = x_train[10000:]
y_val = y_train[:10000]
part_y_train = y_train[10000:]

In [12]:
from keras import models
from keras import layers
from keras import optimizers

model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,))) #one hidden layer
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
print(results)

Epoch 1/4
25000/25000 [==============================] - 9s 342us/step - loss: 0.4506 - acc: 0.8292
Epoch 2/4
25000/25000 [==============================] - 6s 228us/step - loss: 0.2769 - acc: 0.9070
Epoch 3/4
25000/25000 [==============================] - 5s 201us/step - loss: 0.2189 - acc: 0.9249
Epoch 4/4
25000/25000 [==============================] - 10s 415us/step
[0.28889801345825195, 0.88328]


In [13]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,))) #three hidden layers
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
print(results)

Epoch 1/4
25000/25000 [==============================] - 7s 267us/step - loss: 0.4582 - acc: 0.8196
Epoch 2/4
25000/25000 [==============================] - 5s 215us/step - loss: 0.2508 - acc: 0.9066
Epoch 3/4
25000/25000 [==============================] - 5s 210us/step - loss: 0.1951 - acc: 0.9293
Epoch 4/4
25000/25000 [==============================] - 5s 219us/step
[0.3069689905548096, 0.88232]


## Вывод:

1 слой: [0.28889801345825195, 0.88328]

2 слоя: [0.2992042181968689, 0.88172] (результаты взяты из Lecture2)

3 слоя: [0.3069689905548096, 0.88232]

Однослойная модель показывает чуть бОльшую точность, но этот отрыв может исчезнуть при перезапуске.

## Try using layers with more hidden units or fewer hidden units: 32 units, 64 units, and so on.

In [14]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)
print(results)

Epoch 1/4
25000/25000 [==============================] - 6s 258us/step - loss: 0.4244 - acc: 0.8220
Epoch 2/4
25000/25000 [==============================] - 6s 220us/step - loss: 0.2404 - acc: 0.9124
Epoch 3/4
25000/25000 [==============================] - 5s 219us/step - loss: 0.1893 - acc: 0.9300
Epoch 4/4
25000/25000 [==============================] - 6s 228us/step
[0.31684351998806, 0.87772]


In [15]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)
print(results)

Epoch 1/4
25000/25000 [==============================] - 7s 284us/step - loss: 0.4183 - acc: 0.8166
Epoch 2/4
25000/25000 [==============================] - 8s 329us/step - loss: 0.2336 - acc: 0.9105
Epoch 3/4
25000/25000 [==============================] - 7s 289us/step - loss: 0.1841 - acc: 0.9287
Epoch 4/4
25000/25000 [==============================] - 7s 268us/step
[0.3457731051063538, 0.87328]


## Вывод
16 нейронов: [0.2992042181968689, 0.88172] (результаты взяты из Lecture2)

32 нейрона: [0.31684351998806, 0.87772]

64 нейрона: [0.3457731051063538, 0.87328]

Модель с 16 нейронами дает большую точность.

## Try using the mse loss function instead of binary_crossentropy.

In [16]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='mse',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)
print(results)

Epoch 1/4
25000/25000 [==============================] - 5s 214us/step - loss: 0.1483 - acc: 0.8120
Epoch 2/4
25000/25000 [==============================] - 5s 192us/step - loss: 0.0761 - acc: 0.9117
Epoch 3/4
25000/25000 [==============================] - 5s 210us/step - loss: 0.0577 - acc: 0.9310
Epoch 4/4
25000/25000 [==============================] - 5s 214us/step
[0.09209134727954864, 0.87588]


## Вывод

Сравниваем функции потерь:

binary_crossentropy [0.2992042181968689, 0.88172] (результаты взяты из Lecture2)

mse [0.09209134727954864, 0.87588]

У нас задача классификации, в ней используется binary_crossentropy. MSE для задач регрессии. Но при MSE заметно уменьшается test loss.

## Try using the tanh activation (an activation that was popular in the early days of neural networks) instead of relu.

In [17]:
model = models.Sequential()
model.add(layers.Dense(16, activation='tanh', input_shape=(10000,)))
model.add(layers.Dense(16, activation='tanh'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)
print(results)

Epoch 1/4
25000/25000 [==============================] - 6s 252us/step - loss: 0.4290 - acc: 0.8332
Epoch 2/4
25000/25000 [==============================] - 5s 187us/step - loss: 0.2367 - acc: 0.9141
Epoch 3/4
25000/25000 [==============================] - 5s 206us/step - loss: 0.1769 - acc: 0.9353
Epoch 4/4
25000/25000 [==============================] - 6s 230us/step
[0.35529177773714066, 0.86956]


## Вывод

Сравниваем функции активации:
    
relu:[0.2992042181968689, 0.88172] (результаты взяты из Lecture2)
    
tanh:[0.35529177773714066, 0.86956]
    
При использовании функции активации "tanh" точность уменьшается, а test loss увеличивается.

----------------------------

## Part 2
## Try using larger or smaller layers: 32 units, 128 units, and so on.

In [1]:
from keras.datasets import reuters

(train_data, train_labels), (test_data, test_labels) = reuters.load_data(
    num_words=10000)

Using TensorFlow backend.


In [2]:
import numpy as np

def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.
    return results

x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

In [3]:
from keras.utils.np_utils import to_categorical

one_hot_train_labels = to_categorical(train_labels)
one_hot_test_labels = to_categorical(test_labels)

In [4]:
x_val = x_train[:1000]
partial_x_train = x_train[1000:]

y_val = one_hot_train_labels[:1000]
partial_y_train = one_hot_train_labels[1000:]

In [5]:
from keras import models
from keras import layers
from keras import optimizers

model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 352us/step - loss: 3.0949 - acc: 0.4847 - val_loss: 2.3680 - val_acc: 0.5730
Epoch 2/9
7982/7982 [==============================] - 2s 196us/step - loss: 1.9645 - acc: 0.6055 - val_loss: 1.7334 - val_acc: 0.6240
Epoch 3/9
7982/7982 [==============================] - 2s 195us/step - loss: 1.4699 - acc: 0.6863 - val_loss: 1.4556 - val_acc: 0.6850
Epoch 4/9
7982/7982 [==============================] - 2s 207us/step - loss: 1.2077 - acc: 0.7403 - val_loss: 1.2995 - val_acc: 0.7020
Epoch 5/9
7982/7982 [==============================] - 2s 195us/step - loss: 1.0267 - acc: 0.7751 - val_loss: 1.1925 - val_acc: 0.7360
Epoch 6/9
7982/7982 [==============================] - 2s 195us/step - loss: 0.8850 - acc: 0.8087 - val_loss: 1.1186 - val_acc: 0.7680
Epoch 7/9
7982/7982 [==============================] - 2s 197us/step - loss: 0.7725 - acc: 0.8312 - val_loss: 1.0764 - val_acc: 0.7720
Epoch 8

[1.0746706011883405, 0.7573463936416782]

In [6]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 408us/step - loss: 2.6565 - acc: 0.5316 - val_loss: 1.7434 - val_acc: 0.6420
Epoch 2/9
7982/7982 [==============================] - 2s 253us/step - loss: 1.4065 - acc: 0.7146 - val_loss: 1.3234 - val_acc: 0.7180
Epoch 3/9
7982/7982 [==============================] - 2s 245us/step - loss: 1.0457 - acc: 0.7760 - val_loss: 1.1490 - val_acc: 0.7450
Epoch 4/9
7982/7982 [==============================] - 2s 247us/step - loss: 0.8323 - acc: 0.8218 - val_loss: 1.0537 - val_acc: 0.7820
Epoch 5/9
7982/7982 [==============================] - 2s 249us/step - loss: 0.6693 - acc: 0.8583 - val_loss: 0.9862 - val_acc: 0.7900
Epoch 6/9
7982/7982 [==============================] - 2s 249us/step - loss: 0.5375 - acc: 0.8864 - val_loss: 0.9375 - val_acc: 0.8080
Epoch 7/9
7982/7982 [==============================] - 2s 252us/step - loss: 0.4369 - acc: 0.9085 - val_loss: 0.9857 - val_acc: 0.7850
Epoch 8

[0.9841250855266677, 0.7867319679430098]

In [6]:
from keras import models
from keras import layers
from keras import optimizers

model = models.Sequential()
model.add(layers.Dense(128, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(46, activation='softmax')) 

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 4s 491us/step - loss: 2.2444 - acc: 0.5604 - val_loss: 1.4596 - val_acc: 0.6720
Epoch 2/9
7982/7982 [==============================] - 3s 352us/step - loss: 1.1564 - acc: 0.7473 - val_loss: 1.1595 - val_acc: 0.7380
Epoch 3/9
7982/7982 [==============================] - 3s 402us/step - loss: 0.8395 - acc: 0.8141 - val_loss: 1.0144 - val_acc: 0.7770
Epoch 4/9
7982/7982 [==============================] - 3s 429us/step - loss: 0.6210 - acc: 0.8673 - val_loss: 0.9314 - val_acc: 0.8060
Epoch 5/9
7982/7982 [==============================] - 3s 421us/step - loss: 0.4475 - acc: 0.9070 - val_loss: 0.8986 - val_acc: 0.8020
Epoch 6/9
7982/7982 [==============================] - 3s 360us/step - loss: 0.3452 - acc: 0.9236 - val_loss: 0.8798 - val_acc: 0.8180
Epoch 7/9
7982/7982 [==============================] - 3s 368us/step - loss: 0.2760 - acc: 0.9390 - val_loss: 1.0626 - val_acc: 0.7760
Epoch 8

[1.0182387423324246, 0.7938557435706165]

## Вывод:
32 нейрона: [1.0746706011883405, 0.7573463936416782]

64 нейрона: [0.9841250855266677, 0.7867319679430098]

128 нейронов:[1.0182387423324246, 0.7938557435706165]

Если мы меняем размерность слоев, на 128 нейронах точность модели оказывается лучше. Однако, в предыдущем запуске модель с 64 нейронами показала бОльшую точность.

## You used two hidden layers. Now try using a single hidden layer, or three hidden layers.

In [7]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,))) #one hidden layer
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 338us/step - loss: 2.4983 - acc: 0.5727 - val_loss: 1.7686 - val_acc: 0.6660
Epoch 2/9
7982/7982 [==============================] - 2s 262us/step - loss: 1.4390 - acc: 0.7253 - val_loss: 1.3275 - val_acc: 0.7250
Epoch 3/9
7982/7982 [==============================] - 2s 245us/step - loss: 1.0475 - acc: 0.7884 - val_loss: 1.1264 - val_acc: 0.7700
Epoch 4/9
7982/7982 [==============================] - 2s 252us/step - loss: 0.8164 - acc: 0.8363 - val_loss: 1.0147 - val_acc: 0.8000
Epoch 5/9
7982/7982 [==============================] - 2s 252us/step - loss: 0.6541 - acc: 0.8735 - val_loss: 0.9383 - val_acc: 0.8070
Epoch 6/9
7982/7982 [==============================] - 2s 253us/step - loss: 0.5338 - acc: 0.8971 - val_loss: 0.8914 - val_acc: 0.8150
Epoch 7/9
7982/7982 [==============================] - 2s 249us/step - loss: 0.4402 - acc: 0.9134 - val_loss: 0.8639 - val_acc: 0.8170
Epoch 8

[0.9057835927523361, 0.7996438112199465]

In [8]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 396us/step - loss: 2.5588 - acc: 0.5589 - val_loss: 1.6607 - val_acc: 0.6580
Epoch 2/9
7982/7982 [==============================] - 2s 252us/step - loss: 1.3588 - acc: 0.7196 - val_loss: 1.2472 - val_acc: 0.7130
Epoch 3/9
7982/7982 [==============================] - 2s 247us/step - loss: 1.0078 - acc: 0.7855 - val_loss: 1.0997 - val_acc: 0.7720
Epoch 4/9
7982/7982 [==============================] - 2s 253us/step - loss: 0.7979 - acc: 0.8287 - val_loss: 1.0073 - val_acc: 0.7820
Epoch 5/9
7982/7982 [==============================] - 2s 250us/step - loss: 0.6428 - acc: 0.8607 - val_loss: 0.9699 - val_acc: 0.7870
Epoch 6/9
7982/7982 [==============================] - 2s 310us/step - loss: 0.5138 - acc: 0.8896 - val_loss: 0.9088 - val_acc: 0.8090
Epoch 7/9
7982/7982 [==============================] - 2s 263us/step - loss: 0.4193 - acc: 0.9117 - val_loss: 0.8806 - val_acc: 0.8220
Epoch 8

[0.9845317970506867, 0.7885129118698151]

In [9]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,))) #three hidden layers
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 392us/step - loss: 2.5556 - acc: 0.5046 - val_loss: 1.6809 - val_acc: 0.6500
Epoch 2/9
7982/7982 [==============================] - 2s 244us/step - loss: 1.4205 - acc: 0.6987 - val_loss: 1.3512 - val_acc: 0.6910
Epoch 3/9
7982/7982 [==============================] - 2s 282us/step - loss: 1.1002 - acc: 0.7508 - val_loss: 1.1951 - val_acc: 0.7250
Epoch 4/9
7982/7982 [==============================] - 2s 296us/step - loss: 0.8765 - acc: 0.8042 - val_loss: 1.0903 - val_acc: 0.7570
Epoch 5/9
7982/7982 [==============================] - 2s 284us/step - loss: 0.6928 - acc: 0.8470 - val_loss: 1.0250 - val_acc: 0.7950
Epoch 6/9
7982/7982 [==============================] - 2s 270us/step - loss: 0.5530 - acc: 0.8800 - val_loss: 1.0816 - val_acc: 0.7750
Epoch 7/9
7982/7982 [==============================] - 2s 273us/step - loss: 0.4461 - acc: 0.9027 - val_loss: 0.9923 - val_acc: 0.7970
Epoch 8

[1.1366345856939164, 0.7675868210682143]

## Вывод:

1 слой: [0.9057835927523361, 0.7996438112199465]

2 слоя: [0.9845317970506867, 0.7885129118698151]

3 слоя: [1.1366345856939164, 0.7675868210682143]

Если мы меняем количество слоев, лучшей оказывается однослойная модель (точность почти 0,8 и test loss наименьшая). Наши данные несложные, на 3 слоях модель переобучается.